internal/api/v1/proxy/proxy.go

```diff
@@ -0,0 +1,126 @@
+// Copyright © 2021 - 2023 SUSE LLC
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/epinio/epinio/helpers/kubernetes"
+	"github.com/epinio/epinio/internal/cli/server/requestctx"
+	apierror "github.com/epinio/epinio/pkg/api/core/v1/errors"
+	"github.com/gin-gonic/gin"
+	"k8s.io/client-go/rest"
+)
+
+func RunProxy(ctx context.Context, rw http.ResponseWriter, req *http.Request, destination *url.URL) apierror.APIErrors {
+	clientSetHTTP1, err := kubernetes.GetHTTP1Client(ctx)
+	if err != nil {
+		return apierror.InternalError(err)
+	}
+
+	httpClient := clientSetHTTP1.CoreV1().RESTClient().(*rest.RESTClient).Client
+
+	p := httputil.ReverseProxy{
+		Director: func(req *http.Request) {
+			req.URL = destination
+			req.Host = destination.Host
+			// let kube authentication work
+			delete(req.Header, "Cookie")
+			delete(req.Header, "Authorization")
+		},
+		Transport:     httpClient.Transport,
+		FlushInterval: time.Millisecond * 100,
+	}
+
+	p.ServeHTTP(rw, req)
+
+	return nil
+}
+
+type TCPProxy struct {
+	GCtx               *gin.Context
+	Address            string
+	StopChan           <-chan struct{}
+	UpgradedConnection net.Conn
+	ProxyConnection    net.Conn
+}
+
+func NewTCPProxy(gctx *gin.Context, upgradedConnection net.Conn, address string, stopChan <-chan struct{}) (*TCPProxy, error) {
+	if !strings.Contains(address, ":") {
+		address += ":80"
+	}
+	return &TCPProxy{
+		GCtx:               gctx,
+		UpgradedConnection: upgradedConnection,
+		Address:            address,
+		StopChan:           stopChan,
+	}, nil
+}
+
+func (p *TCPProxy) Start() error {
+	ctx := p.GCtx.Request.Context()
+	logger := requestctx.Logger(ctx).WithName("PortForward")
+	var d net.Dialer
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	var err error
+	if p.ProxyConnection, err = d.DialContext(ctx, "tcp", p.Address); err != nil {
+		logger.Error(err, "dialing service")
+		return err
+	}
+
+	return p.handleConnections()
+}
```

Comment:
It's better to give the possibility to tweak the Timeout from the caller, and probably also the stop chan could be an argument of this method.

```go
func (p *TCPProxy) Start(ctx context.Context, stopChan <-chan struct{}) error {}
```

With this approach you can stop and start again the Proxy, if needed. Otherwise the closed chan will remain closed, and you need to create a new Proxy in order to Start it again.

Reply:
looking better at the code, `stopChan` was actually not used in this context, so I got rid of it.